diff --git a/Controller/Onepage/Success.php b/Controller/Onepage/Success.php
index 7df11761b..2bbb4b285 100644
--- a/Controller/Onepage/Success.php
+++ b/Controller/Onepage/Success.php
@@ -24,7 +24,10 @@ class Success extends \Magento\Checkout\Controller\Onepage
         $resultPage = $this->resultPageFactory->create();
         $this->_eventManager->dispatch(
             'checkout_onepage_controller_success_action',
-            ['order_ids' => [$session->getLastOrderId()]]
+            [
+                'order_ids' => [$session->getLastOrderId()],
+                'order' => $session->getLastRealOrder()
+            ]
         );
         return $resultPage;
     }
