diff --git a/Block/Address/Edit.php b/Block/Address/Edit.php
index b1ac4e83..7d1b782f 100644
--- a/Block/Address/Edit.php
+++ b/Block/Address/Edit.php
@@ -127,7 +127,7 @@ class Edit extends \Magento\Directory\Block\Data
 
         if ($postedData = $this->_customerSession->getAddressFormData(true)) {
             $postedData['region'] = [
-                'region_id' => $postedData['region_id'],
+                'region_id' => isset($postedData['region_id']) ? $postedData['region_id'] : null,
                 'region' => $postedData['region'],
             ];
             $this->dataObjectHelper->populateWithArray(
